doc/design/python_api.md

```diff
@@ -0,0 +1,186 @@
+
+`Program` is the description of the whole training process and there can only be one `Program` object, which is created automatically by the system at the very beginning. `Program` is formed by a series of `Block`.
+
+```python
+class Program(objects):
+    def __init__(self):
+        self.blocks = vector<Block>()
+        self.blocks.append(Block(None))
+        self.current_block_idx = 0
+
+    def get_block(block_idx):
+        return self.blocks[block_idx]
+
+    def current_block():
+        return self.get_block(self.current_block_idx)
+    
+    def fallback_current_block():
+        self.current_block_idx = self.current_block().parent_idx
+
+
+    def create_block():
+        new_block_idx = len(self.block)
+        self.blocks.append(Block(parent_idx=self.current_block_idx,
+                                 idx=new_block_idx))
+        self.current_block_idx = new_block_idx
+```
+
+`Program` will create the first block in its constructor. The first block is called 'global block'. It is where all parameters are stored.
+
+### Block
+
+Block is just like programming languages `{}`, which contains many operators and variables. There are two data fields in `Block`.  1) An associate map, whose key is variable name and value is variable itself; 2) A list of operators.
+
+The block is hierarchical because PaddlePaddle supports RNN and IfElse. For example, RNN is like `for-loop` in programming languages. There is new `block` inside a `for-loop`. To represent hierarchies, `Block` stores the index of `parent Block` inside. The 'index' means the block's position in `Program`'s `blocks`. If `parent_idx=None`, the block itself is the outermost block, i.e., the 'global block'.
+
+
+```python
+class Block(objects):
+    def __init__(self, parent_idx, idx):
+        self.vars = map<string, Variable>()
+        self.ops = vector<Operator>()
+        self.idx = idx
+        self.parent_idx = parent_idx
+    
+    def create_var(self, ...):
+        # create variable in `self.vars`
+        return Variable(...)
+    
+    
+    def create_global_var(self, ...):
+        if self.parent_idx is not None:
+            parent_block = program.get_block(parent_idx)
+            return parent_block.create_global_var(...)
+        else:
+            return self.create_var(...)
+    
+    def create_parameter(self, ...):
+        return self.create_global_var(...)
+    
+    def append_operator(self, ...):
+        self.ops.append(...)
+        
+    def prepend_operator(self, ...):
+       self.ops.prepend(...)
+```
+
+Users are able to create a global variable inside any block since they many create parameters inside a RNN or IfElse. All parameters should be stored in the global block, not the step block in RNN.
+
+Users can create local variables for outputs of operators. Users can also append and prepend an operator in current block. Prepending `random initialize` operator or `load` operator is very useful to initialize parameters before training.
```

Comment:
"current block" is mentioned 3 times in this document, what exactly does it mean. Is it a global variable?